File: examples/output-handlers/output-handler.py
"""
Example Python output handler for promptfoo evaluation results.
This demonstrates processing evaluation results in Python.
"""

import json
from datetime import datetime


def handle_output(data):
    """Process evaluation results and generate custom reports."""
    print("\n=== Python Evaluation Results Handler ===")
    print(f"Eval ID: {data['evalId']}")
    print(f"Timestamp: {datetime.now().isoformat()}")

    stats = data["results"]["stats"]
    success_rate = (
        (stats["successes"] / stats["total"] * 100) if stats["total"] > 0 else 0
    )

    print(
        f"\nOverall Results: {stats['successes']}/{stats['total']} passed ({success_rate:.2f}%)"
    )
    print(f"Failures: {stats['failures']}")
    print(f"Tokens used: {stats['tokenUsage']['total']}")

    # Check for failures
    if stats["failures"] > 0:
        print("\n⚠️  Failures detected:")

        for idx, result in enumerate(data["results"]["results"]):
            if not result["success"]:
                print(f"\n  Test {idx + 1}:")
                print(f"    Prompt: {result['prompt']['raw']}")
                print(f"    Variables: {json.dumps(result['vars'])}")
                error_msg = result.get("error", "Assertion failed")
                print(f"    Reason: {error_msg}")

    # Example: Export to CSV
    # import csv
    # with open('results.csv', 'w') as f:
    #     writer = csv.writer(f)
    #     writer.writerow(['Test', 'Success', 'Prompt', 'Variables'])
    #     for idx, result in enumerate(data['results']['results']):
    #         writer.writerow([
    #             idx + 1,
    #             result['success'],
    #             result['prompt']['raw'],
    #             json.dumps(result['vars'])
    #         ])

    # Example: Send to webhook
    # import requests
    # webhook_data = {
    #     'evalId': data['evalId'],
    #     'successRate': success_rate,
    #     'failures': stats['failures']
    # }
    # requests.post('https://example.com/webhook', json=webhook_data)

    # Create summary
    summary = {
        "evalId": data["evalId"],
        "timestamp": datetime.now().isoformat(),
        "successRate": round(success_rate, 2),
        "totalTests": stats["total"],
        "failures": stats["failures"],
        "tokenUsage": stats["tokenUsage"]["total"],
    }

    print(f"\n📊 Summary: {json.dumps(summary, indent=2)}")

    return summary


# Alternative function with different processing
def process_for_dashboard(data):
    """Process results specifically for dashboard display."""
    results = []
    for result in data["results"]["results"]:
        results.append(
            {
                "success": result["success"],
                "topic": result["vars"].get("topic", "unknown"),
                "response": result.get("response", {}).get("output")
                if result["success"]
                else None,
                "latency": result["latencyMs"],
            }
        )

    results_list = data["results"]["results"]
    avg_latency = (
        sum(r["latencyMs"] for r in results_list) / len(results_list)
        if results_list
        else 0
    )

    return {"evalId": data["evalId"], "results": results, "avgLatency": avg_latency}

File: examples/output-handlers/output-handler.js
/**
 * Example output handler for promptfoo evaluation results
 * This handler receives the evaluation results and can process them
 * for custom reporting, alerting, or integration with other systems.
 */

module.exports = async function handleEvaluationResults(data) {
  console.log('\n=== Evaluation Results Handler ===');
  console.log(`Eval ID: ${data.evalId}`);
  console.log(`Timestamp: ${new Date().toISOString()}`);

  const { stats } = data.results;
  const successRate = ((stats.successes / stats.total) * 100).toFixed(2);

  console.log(`\nOverall Results: ${stats.successes}/${stats.total} passed (${successRate}%)`);
  console.log(`Failures: ${stats.failures}`);
  console.log(`Tokens used: ${stats.tokenUsage.total}`);

  // Check for failures and log them
  if (stats.failures > 0) {
    console.log('\n⚠️  Failures detected:');

    data.results.results.forEach((result, idx) => {
      if (!result.success) {
        console.log(`\n  Test ${idx + 1}:`);
        console.log(`    Prompt: ${result.prompt.raw}`);
        console.log(`    Variables: ${JSON.stringify(result.vars)}`);
        console.log(`    Reason: ${result.error || 'Assertion failed'}`);
      }
    });
  }

  // Example: Send metrics to monitoring system
  // await sendToMonitoring({
  //   metric: 'promptfoo.eval.success_rate',
  //   value: successRate,
  //   tags: { evalId: data.evalId }
  // });

  // Example: Trigger alerts for low success rates
  if (successRate < 80) {
    console.log(`\n🚨 Alert: Success rate below 80% threshold!`);
    // await sendAlert(`Evaluation ${data.evalId} has low success rate: ${successRate}%`);
  }

  // Example: Save summary to database
  const summary = {
    evalId: data.evalId,
    timestamp: new Date().toISOString(),
    successRate: Number.parseFloat(successRate),
    totalTests: stats.total,
    failures: stats.failures,
    tokenUsage: stats.tokenUsage.total,
  };

  console.log('\n📊 Summary:', JSON.stringify(summary, null, 2));

  // Return value is optional but can be used for chaining
  return summary;
};
